pkg/kubelet/kubelet_pods.go

```diff
@@ -1734,26 +1739,28 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 	}
 
 	for _, container := range containers {
+		klog.V(2).InfoS("setting old container status", "pod", klog.KObj(pod), "container", container)
 		status := &v1.ContainerStatus{
 			Name:  container.Name,
 			Image: container.Image,
 			State: defaultWaitingState,
 		}
 		oldStatus, found := oldStatuses[container.Name]
 		if found {
-			if oldStatus.State.Terminated != nil {
+			if isInitContainer && oldStatus.State.Terminated != nil {
```

Comment:
I think this will ensure the old status still gets set while avoiding the bugs fixed by https://github.com/kubernetes/kubernetes/pull/54593 and https://github.com/kubernetes/kubernetes/pull/51644